crates/sui/tests/protocol_version_tests.rs

```diff
@@ -676,8 +676,8 @@ mod sim_only_tests {
         // The system state object will be upgraded next time we execute advance_epoch transaction
```

Comment:
Since the new framework replaces the old one, the override on line 668 is a noop? so we didn't really do any framework upgrade. System state obj is upgraded though since genesis created version 1.

Reply:
In general I think it's useful to keep the "old" framework somewhere (and without modifying it) as we make changes to the new framework. And the test here just needs to start / genesis with the old framework via `test_cluster_builder.with_objects(old_framework)`. Then at the next epoch change, we will automatically pick up the new framework without needing the `sui_system_injection` override, and we check that the upgrade is compatible and can go through successfully. The old framework can be deleted after the new one has been deployed, and the cycle starts again. What do you think?

Reply:
For 100% coverage, we should keep a binary representation of the live system packages around, or do something similar to the TS devnet E2E tests and add a CI job to checkout our release tracking branches (`testnet`, `devnet`), build all system packages there and then come back to main and confirm that the version of the framework in main is compatible with all of those.

It's on my queue to add this, but if you're interested in it and have cycles you're welcome to it!

Reply:
What I did was initially putting the new sui-system in a new directory, and run this test by pointing the override to that new directory. After test succeeds, I then moved the new framework to override the old framework

Reply:
I can certainly keep a copy of the old framework

Reply:
We should not keep around a copy of the old framework source, we need to keep around a copy of the old framework bytecode, because otherwise we risk missing a compatibility issue due to a framework change.

Reply:
We probably should also put the bytecode version when testnet branch was cut

Reply:
Yeah that's right -- we should take/update a snapshot every time we cut a release to go out to devnet/testnet/mainnet, and the test should make sure the current framework source, when built is compatible with all live versions.

In terms of trust, I'm not too worried -- I think we should make the process of taking a snapshot as easy as running a script, but otherwise yes, it's fine to trust, because the upgrade process itself prevents incompatible changes going in, and what we are talking about here is just an early warning system to prevent landing incompatible changes and only realising when we try and run a protocol upgrade and it doesn't go through.